internal/fieldmask/fieldmask.go

```diff
@@ -0,0 +1,165 @@
+package fieldmask
+
+import (
+	"context"
+	"net/http"
+	"strings"
+
+	jsoniter "github.com/json-iterator/go"
+	"github.com/tidwall/gjson"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/types/known/fieldmaskpb"
+)
+
+const key = "fields"
+
+// FieldMask is recursive structure to define a path mask
+type FieldMask map[string]FieldMask
+
+// New create a FieldMask from the input array of paths. The array should contain JSON paths with dit "." notation.
+func New(paths []string) FieldMask {
+	mask := make(FieldMask)
+	for _, path := range paths {
+		current := mask
+		fields := strings.Split(path, ".")
+		for _, field := range fields {
+			c, ok := current[field]
+			if !ok {
+				c = make(FieldMask)
+				current[field] = c
+			}
+			current = c
+		}
+	}
+	return mask
+}
+
+// Filter takes a Proto message as input and updates the message according to the FieldMask.
+func (fm FieldMask) Filter(message proto.Message) {
+	if len(fm) == 0 {
+		return
+	}
+
+	reflect := message.ProtoReflect()
+	reflect.Range(func(fd protoreflect.FieldDescriptor, v protoreflect.Value) bool {
+		mask, ok := fm[string(fd.Name())]
+		if !ok {
+			reflect.Clear(fd)
+		}
+
+		if len(mask) == 0 {
+			return true
+		}
+
+		switch {
+		case fd.IsMap():
+			m := reflect.Get(fd).Map()
+			m.Range(func(k protoreflect.MapKey, v protoreflect.Value) bool {
+				if fm, ok := mask[k.String()]; ok {
+					if i, ok := v.Interface().(protoreflect.Message); ok && len(fm) > 0 {
+						fm.Filter(i.Interface())
+					}
+				} else {
+					m.Clear(k)
+				}
+				return true
+			})
+		case fd.IsList():
+			list := reflect.Get(fd).List()
+			for i := 0; i < list.Len(); i++ {
+				mask.Filter(list.Get(i).Message().Interface())
+			}
+		case fd.Kind() == protoreflect.MessageKind:
+			mask.Filter(reflect.Get(fd).Message().Interface())
+		case fd.Kind() == protoreflect.BytesKind:
```

Comment:
so it appears `BytesKind` is where the recursion will short circuit and return, and that map,list,message cases will all eventually boil down to bytes kind, correct?

Reply:
In our implementation, most of the cases will boil down to `BytesKind`, but not all. In some messages, there are other types of fields also. But this is more of a generic implementation, in case in future we include child proto or `Lists` in our messages.

Reply:
ack thanks for confirming